changes/2020-06-04_how-to-fail-with-keyrings/background.md

```diff
@@ -0,0 +1,166 @@
+
+# Requirements
+
+In order to resolve these inconsistencies,
+we need to define what kinds of failure modes we allow,
+what causes them,
+and how the ESDK framework components handle them.
+
+# Success Measurements
+
+We will know we are succeeding when we have consensus on
+the definition of when keyrings fail,
+what causes those failures,
+and how ESDK framework components handle those failures.
+
+# Out of Scope
+
+- The specifics of how ESDK framework components
+  communicate failure information between themselves
+  is out of scope for this document.
+  That will be addressed separately once we have
+  consensus on this document.
+
+- The specific nature of _how_ ESDK framework components indicate failure
+  is out of scope for this document.
+  That will depend on the best practices for each language.
+
+- We need to define detailed decryption contracts
+  (defined in [Issues and Alternatives](#issues-and-alternatives))
+  for each keyring defined in the ESDK specification.
+  This will be addressed separately once we have
+  consensus on this document.
+
+- A mechanism to interrogate a keyring to determine
+  the requirements of its decryption contract
+  (defined in [Issues and Alternatives](#issues-and-alternatives))
+  is out of scope for this document.
+  We should pursue this once decryption contracts are defined
+  for each keyring.
+
+# Issues and Alternatives
+
+## Issue 0 : What does a keyring do?
+
+In order to define when a keyring fails,
+we must first clearly define what a keyring does.
+When a keyring encrypts a data key,
+it creates one or more artifacts that define
+a set of requirements that MUST be met
+in order to use that artifact
+to obtain the data key.
+We call this set of requirements the "decryption contract".
+In order to obtain the data key from the artifact,
+a keyring MUST "satisfy" the decryption contract.
+
+For example, some decryption contracts for keyrings include:
+
+- Raw RSA keyring
+
+  - The keyring MUST have access to the private wrapping key.
+
+- Raw AES keyring
+
+  - The keyring MUST have access to the symmetric wrapping key.
+  - The encryption context MUST exactly match
+    the encryption context used on encrypt.
+
+    - NOTE: In compliant implementations this is a MUST,
+      but mathematically this is a SHOULD
+      because it is possible to create a keyring
+      that ignores this clause in the decryption contract
+      by using AES-CTR rather than AES-GCM.
+
+- AWS KMS keyring (single CMK)
+
+  - The AWS KMS client MUST be configured with credentials for
+    an AWS principal that will pass a `kms:Decrypt` authorization check
+    for the encrypted data key.
+  - The encryption context MUST exactly match
+    the encryption context used on encrypt.
+
+- Multi-keyring:
+
+  - Satisfying any member keyring's decryption contract
+    MUST be sufficient to satisfy
+    the multi-keyring's decryption contract.
+
+## Issue 1 : When MUST a keyring fail?
```

Comment:
Is it reasonable to consider keyrings which are only one sided? Maybe many-to-one?

Many different key rings only "encrypt" but the resulting artifacts can all be decrypted by a single standardized keyring?